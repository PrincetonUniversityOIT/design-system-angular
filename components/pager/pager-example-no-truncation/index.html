
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Pager No Truncation Example - Princeton Design System</title>
   <link rel="stylesheet" href="/design-system-angular/jazz_custom.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_fonts.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_curated_icons.css">
   <link rel="stylesheet" href="/design-system-angular/prism-atom-dark.css">
</head>
<body>
<!-- Example Start -->
<div style="width:90%;height:90%;padding:10px;">

<jazz-pager-component current-page="4" total-pages="7"></jazz-pager-component>


</div>
<!-- Example End -->
<script src="/design-system-angular/runtime.js"></script>
<script src="/design-system-angular/polyfills.js"></script>
<script src="/design-system-angular/main.js"></script>
</body>
</html>
