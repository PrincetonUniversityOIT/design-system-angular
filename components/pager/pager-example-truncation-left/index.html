
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Pager Truncation on Left Side Example - Princeton Design System</title>
   <link rel="stylesheet" href="/design-system-angular/jazz_custom.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_fonts.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_curated_icons.css">
   <link rel="stylesheet" href="/design-system-angular/prism-atom-dark.css">
</head>
<body>
<!-- Example Start -->
<div style="width:90%;height:90%;padding:10px;">

<jazz-pager-component current-page="19" total-pages="20"></jazz-pager-component>


</div>
<!-- Example End -->
</body>
</html>
