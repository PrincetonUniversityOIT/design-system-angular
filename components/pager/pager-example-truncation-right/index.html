
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Pager Truncation on Right Side Example - Princeton Design System</title>
   <link rel="stylesheet" href="/design-system-angular/jazz_custom.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_fonts.css">
   <link rel="stylesheet" href="/design-system-angular/jazz_curated_icons.css">
   <link rel="stylesheet" href="/design-system-angular/prism-atom-dark.css">
</head>
<body>
<!-- Example Start -->
<div style="width:90%;height:90%;padding:10px;">

<jazz-pager-component current-page="2" total-pages="20"></jazz-pager-component>


</div>
<!-- Example End -->
<script src="/design-system-angular/runtime.js"></script>
<script src="/design-system-angular/polyfills.js"></script>
<script src="/design-system-angular/main.js"></script>
</body>
</html>
