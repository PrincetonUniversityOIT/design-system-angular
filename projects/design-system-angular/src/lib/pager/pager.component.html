<nav *ngIf="totalPages && currentPage" class="jazz-pager" role="navigation" aria-label="pagination">
  <ul>
    <li>
      <a *ngIf="(currentPage !== 1)" href="javascript:void(0);" (click)="setPage(currentPage - 1)" data-page="Previous"><i class="jazz-icon jazz-icon-caret-left"></i>Previous</a>
      <span *ngIf="(currentPage === 1)" [attr.aria-disabled]="true" data-page="Previous"><i class="jazz-icon jazz-icon-caret-left"></i>Previous</span>
    </li>
    <li *ngFor="let page of iterablePages()" attr.aria-current="{{ (currentPage === (page + 1)) ?'page': ''}}">
      <a *ngIf="page >= 0" href="javascript:void(0);" (click)="setPage(page + 1)"
         attr.aria-label="Go to {{page + 1}}"
         attr.data-page="{{page + 1}}">{{page + 1}}
      </a>
      <span *ngIf="page < 0">...</span>
    </li>
    <li>
      <a *ngIf="(currentPage !== totalPages)" href="javascript:void(0);" (click)="setPage(currentPage+ 1)" data-page="Next">Next<i class="jazz-icon jazz-icon-caret-right"></i></a>
      <span *ngIf="(currentPage === totalPages)" [attr.aria-disabled]="true" data-page="Next">Next<i class="jazz-icon jazz-icon-caret-right"></i></span>
    </li>
  </ul>
</nav>
