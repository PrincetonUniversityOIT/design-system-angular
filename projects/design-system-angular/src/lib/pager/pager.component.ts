import {Component, EventEmitter, Input, Output} from '@angular/core';

/**
 * The Pager allows users to navigate through a set of items or rows which have been separated into pages.
 *
 * @example
  `` `
  <jazz-pager [totalPages]="20" [currentPage]="5" (pagerChange)="changePage($event);"></jazz-pager>
   `` `
 */
@Component({
  // tslint:disable-next-line:component-selector
  selector: 'jazz-pager',
  templateUrl: './pager.component.html'
})
export class PagerComponent {

  @Input('totalPages')
  public totalPages: number;

  @Input('currentPage')
  public currentPage: number;

  @Output()
  public pagerChange: EventEmitter<any> = new EventEmitter();

  constructor() {}

  iterablePages(): number[] {

    const pages: number[] = [];
    const delta = 4;
    let maxPages = 9;
    let truncate = true;
    let pageNum = 0;

    if (this.totalPages < maxPages) {
      maxPages = this.totalPages;
      truncate = false;
      pageNum = 0;
    } else {
      if (this.currentPage - delta < 0) {
        pageNum = 0;
      } else if (this.currentPage + delta > this.totalPages - 1) {
        pageNum = this.totalPages - maxPages;
      } else {
        pageNum = this.currentPage - delta;
      }
    }

    for (let pageIdx = 0; pageIdx < maxPages; pageIdx++) {
      if (truncate) {
        if (pageIdx === 0) {
          // always show the first page number
          pages.push(0);
        } else if (pageIdx === 1 && pageNum !== 1) {
          // show '...' if second page is not a 2
          pages.push(-1);
        } else if (pageIdx === maxPages - 1) {
          // always show the last page number
          pages.push(this.totalPages - 1);
        } else if (pageIdx === maxPages - 2 && pageNum !== this.totalPages - 2) {
          // show '...' if there is a gap between next to last page and last page
          pages.push(-1);
        } else {
          pages.push(pageNum);
        }
      } else {
        pages.push(pageNum);
      }
      pageNum++;
    }

    return pages;
  }

  /**
   * This is used to set the next page in the pager based on the page that was clicked.
   */
  setPage(page: number): void {
    if (page >= 0 && page <= this.totalPages) {
      this.currentPage = page;
      this.pagerChange.emit(page);
    }
  }
}
