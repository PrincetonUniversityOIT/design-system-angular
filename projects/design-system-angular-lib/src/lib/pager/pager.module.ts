import {NgModule} from '@angular/core';
import {PagerComponent} from './pager.component';

@NgModule({
  declarations: [
    PagerComponent
  ],
  exports: [
    PagerComponent
  ]
})
export class PagerModule {

}
