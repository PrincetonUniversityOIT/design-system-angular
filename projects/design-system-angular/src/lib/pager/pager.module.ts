import {NgModule} from '@angular/core';
import {PagerComponent} from './pager.component';
import {CommonModule} from '@angular/common';

@NgModule({
  imports: [
    CommonModule
  ],
  declarations: [
    PagerComponent
  ],
  exports: [
    PagerComponent
  ]
})
export class PagerModule {

}
