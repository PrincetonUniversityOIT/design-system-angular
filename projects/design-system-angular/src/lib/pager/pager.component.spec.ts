import {PagerComponent} from './pager.component';
import {ComponentFixture, TestBed} from '@angular/core/testing';
import {By} from '@angular/platform-browser';

describe('PagerComponent', () => {
  let component: PagerComponent;
  let fixture: ComponentFixture<PagerComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ PagerComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(PagerComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should create a simple pager', () => {
    component.currentPage = 5;
    component.totalPages = 10;

    expect(component.currentPage).toEqual(5);
    expect(component.iterablePages()).toEqual([0, -1, 3, 4, 5, 6,  7, 8, 9 ]);

    component.setPage(6);
    expect(component.currentPage).toEqual(6);
    expect(component.iterablePages()).toEqual([0, -1, 3, 4, 5, 6,  7, 8, 9 ]);
  });

  it('should create a double truncated pager', () => {

    component.currentPage = 5;
    component.totalPages = 10;

    expect(component.currentPage).toEqual(5);
    expect(component.iterablePages()).toEqual([0, -1, 3, 4, 5, 6,  7, 8, 9 ]);

    component.setPage(6);
    expect(component.currentPage).toEqual(6);
    expect(component.iterablePages()).toEqual([0, -1, 3, 4, 5, 6,  7, 8, 9 ]);
  });

  it('should move page forward after clicking next', () => {
    component.currentPage = 7;
    component.totalPages = 20;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges();

        jest.spyOn(component, 'setPage');

        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const items = pager.children[0].children;

        const link = items[items.length - 1].nativeElement.querySelector('a');
        link.click();

        expect(component.setPage).toHaveBeenCalled();
        expect(component.currentPage).toEqual(8);
        expect(component.iterablePages()).toEqual([ 0, -1,  6,  7, 8, 9, 10, -1, 19]);
      }
    );
  });

  it('should move page backward after clicking previous', () => {
    component.currentPage = 7;
    component.totalPages = 20;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges();

        jest.spyOn(component, 'setPage');

        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const items = pager.children[0].children;

        const link = items[0].nativeElement.querySelector('a');
        link.click();

        expect(component.setPage).toHaveBeenCalled();
        expect(component.currentPage).toEqual(6);
        expect(component.iterablePages()).toEqual([  0, -1,  4,  5, 6, 7,  8, -1, 19]);
      }
    );
  });

  it('should create a simple pager and show the previous page link as disabled if the current page = 1', () => {
    component.currentPage = 1;
    component.totalPages = 10;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges();

        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const items = pager.children[0].children;

        const span = items[0].children[0].nativeElement;
        expect(span.getAttribute('data-page')).toEqual('Previous');
        expect(span.getAttribute('aria-disabled')).toEqual('true');
      }
    );
  });

  it('should create a simple pager and show the next page link as disabled if the current page = max page', () => {
    component.currentPage = 10;
    component.totalPages = 10;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges();

        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const items = pager.children[0].children;

        const span = items[items.length - 1].children[0].nativeElement;
        expect(span.getAttribute('data-page')).toEqual('Next');
        expect(span.getAttribute('aria-disabled')).toEqual('true');
      }
    );
  });


  it('should render properly for simple pager', () => {
    component.currentPage = 5;
    component.totalPages = 10;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges(); // missed
        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const pagerHTML = pager.nativeElement.innerHTML;
        expect(pagerHTML).toMatchSnapshot();
      }
    );
  });

  it('should render properly for a double truncated pager', () => {
    component.currentPage = 7;
    component.totalPages = 20;

    fixture.whenStable().then(
      () => {
        fixture.detectChanges();

        const pager = fixture.debugElement.query(By.css('.jazz-pager'));
        const pagerHTML = pager.nativeElement.innerHTML;
        expect(pagerHTML).toMatchSnapshot();
      }
    );
  });

});
