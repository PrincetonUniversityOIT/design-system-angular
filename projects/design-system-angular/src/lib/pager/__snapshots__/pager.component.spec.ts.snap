// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagerComponent should render properly for a double truncated pager 1`] = `
"<ul><li><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a data-page=\\"Previous\\"><span>Previous</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><!--bindings={
  \\"ng-reflect-ng-for-of\\": \\"0,-1,5,6,7,8,9,-1,19\\"
}--><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 1\\" data-page=\\"1\\"><span>1</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><span>...</span></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 6\\" data-page=\\"6\\"><span>6</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"page\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 7\\" data-page=\\"7\\"><span>7</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 8\\" data-page=\\"8\\"><span>8</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 9\\" data-page=\\"9\\"><span>9</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 10\\" data-page=\\"10\\"><span>10</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><span>...</span></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 20\\" data-page=\\"20\\"><span>20</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a data-page=\\"Next\\"><span>Next</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li></ul>"
`;

exports[`PagerComponent should render properly for simple pager 1`] = `
"<ul><li><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a data-page=\\"Previous\\"><span>Previous</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><!--bindings={
  \\"ng-reflect-ng-for-of\\": \\"0,-1,3,4,5,6,7,8,9\\"
}--><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 1\\" data-page=\\"1\\"><span>1</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><span>...</span></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 4\\" data-page=\\"4\\"><span>4</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"page\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 5\\" data-page=\\"5\\"><span>5</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 6\\" data-page=\\"6\\"><span>6</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 7\\" data-page=\\"7\\"><span>7</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 8\\" data-page=\\"8\\"><span>8</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 9\\" data-page=\\"9\\"><span>9</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li aria-current=\\"\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a href=\\"javascript:void(0);\\" aria-label=\\"Go to 10\\" data-page=\\"10\\"><span>10</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li><li><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><a data-page=\\"Next\\"><span>Next</span></a><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></li></ul>"
`;
